hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java

```diff
@@ -1527,6 +1652,278 @@ public Bytes getSuccessResultFor(final ExpirableTxnRecord.Builder childRecord) {
+			}
+		}
+
+		@Override
+		public void run(final MessageFrame frame) {
+			Objects.requireNonNull(approveOp);
+			/* --- Build the necessary infrastructure to execute the transaction --- */
+			final var accountStore = createAccountStore();
+			final var tokenStore = createTokenStore(accountStore, sideEffectsTracker);
+			final var payerAccount = accountStore.loadAccount(Id.fromGrpcAccount(EntityIdUtils.accountIdFromEvmAddress(senderAddress)));
+
+			/* --- Execute the transaction and capture its results --- */
+			final var precompileAddress = Address.fromHexString(HTS_PRECOMPILED_CONTRACT_ADDRESS);
+
+			if (isFungible) {
+				final var checkResponseCode = allowanceChecks.allowancesValidation(transactionBody.getCryptoApproveAllowance().getCryptoAllowancesList(),
+						transactionBody.getCryptoApproveAllowance().getTokenAllowancesList(),
+						transactionBody.getCryptoApproveAllowance().getNftAllowancesList(),
+						payerAccount,
+						currentView);
+
+				if (!OK.equals(checkResponseCode)) {
+					throw new InvalidTransactionException(checkResponseCode);
+				}
+
+				final var approveAllowanceLogic = approveAllowanceLogicFactory.newApproveAllowanceLogic(
+						accountStore, tokenStore, dynamicProperties);
+				approveAllowanceLogic.approveAllowance(transactionBody.getCryptoApproveAllowance().getCryptoAllowancesList(),
+						transactionBody.getCryptoApproveAllowance().getTokenAllowancesList(),
+						transactionBody.getCryptoApproveAllowance().getNftAllowancesList(),
+						EntityIdUtils.accountIdFromEvmAddress(frame.getSenderAddress()));
+				frame.addLog(getLogForFungibleAdjustAllowance(precompileAddress));
+			} else {
+				if (approveOp.spender().getAccountNum() == 0) {
+					final var deleteAllowanceLogic = deleteAllowanceLogicFactory.newDeleteAllowanceLogic(accountStore, tokenStore);
+					deleteAllowanceLogic.deleteAllowance(transactionBody.getCryptoDeleteAllowance().getNftAllowancesList(),
+							EntityIdUtils.accountIdFromEvmAddress(frame.getSenderAddress()));
+				} else {
+					final var checkResponseCode = allowanceChecks.allowancesValidation(transactionBody.getCryptoApproveAllowance().getCryptoAllowancesList(),
+							transactionBody.getCryptoApproveAllowance().getTokenAllowancesList(),
+							transactionBody.getCryptoApproveAllowance().getNftAllowancesList(),
+							payerAccount,
+							currentView);
+
+					if (!OK.equals(checkResponseCode)) {
+						throw new InvalidTransactionException(checkResponseCode);
+					}
+
+					final var approveAllowanceLogic = approveAllowanceLogicFactory.newApproveAllowanceLogic(
+							accountStore, tokenStore, dynamicProperties);
+					approveAllowanceLogic.approveAllowance(transactionBody.getCryptoApproveAllowance().getCryptoAllowancesList(),
+							transactionBody.getCryptoApproveAllowance().getTokenAllowancesList(),
+							transactionBody.getCryptoApproveAllowance().getNftAllowancesList(),
+							EntityIdUtils.accountIdFromEvmAddress(frame.getSenderAddress()));
```

Comment:
Lines 1755-1770 could be extracted in a method in the ApprovePrecompile. This method. could be used both for ERC20 and ERC721 tokens.